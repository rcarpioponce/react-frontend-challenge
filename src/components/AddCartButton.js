import React, { useContext } from 'react'
import { Button } from 'react-bootstrap'
import Context from '../context/Context'
import { useProductFetch } from '../hooks/useProductFetch'


export const AddCartButton = ({ productId }) => {
    
    const [context, setContext] = useContext(Context)

    const {data:product } = useProductFetch(productId);

    const handleAddCart = () => {
        context.push(product);
        setContext([...context])
    }

    return (
        <Button variant="primary" onClick={ handleAddCart }>ADD CART</Button>
    )
}
