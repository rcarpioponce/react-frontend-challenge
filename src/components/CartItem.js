import React from 'react'
import { Row, Col } from 'react-bootstrap';


export const CartItem = ({title, price}) => {
    return (
        <Row>
            <Col>{title}</Col>
            <Col>$. {price}</Col>
        </Row>
    )
}
