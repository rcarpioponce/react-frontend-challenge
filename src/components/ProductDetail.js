import React from 'react'
import { useParams } from 'react-router';
import { Row, Col, Spinner } from 'react-bootstrap';
import { useProductFetch } from '../hooks/useProductFetch';
import { AddCartButton } from './AddCartButton';

export const ProductDetail = () => {
    const { id }  = useParams();

    const {data: product, loading } = useProductFetch(id);

    const { title, image, category, description } = product


    return (
        <Row>

            {
                loading ?
                <Col lg='12'>
                    <Spinner animation="border" role="status">
                    <span className="sr-only">Loading...</span>
                    </Spinner>
                </Col> 
                :
                <>
                <Col lg='6'>
                    <img className='img-fluid' src={image} alt={title} />
                </Col>
                <Col lg='6'>
                    <h1>{ title }</h1>
                    <p>{ description }</p>
                    <p>Category: { category }</p>
                    <AddCartButton productId={id} />
                </Col>
                </>           
            }

        </Row>
        
    )
}
