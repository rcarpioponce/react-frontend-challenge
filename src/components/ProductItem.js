import React from 'react'
import PropTypes from 'prop-types';
import { Col, Card, Row } from 'react-bootstrap'
import { Link } from "react-router-dom"
import { AddCartButton } from './AddCartButton'
export const ProductItem = ({ id, title, price, image}) => {
    return (
        <Col lg="3" md="3" xs="12" className='mb-4 h-100'>
            <Card style={{ height: '370px' }}>
                <Card.Img className='img-fluid' variant="top" src={image} alt={title} />
                <Card.Body>
                    <Card.Title>
                        <Link to={`/product/${id}`}>{ title }</Link>
                    </Card.Title>
                    <Row>
                        <Col>
                            $. { price }
                        </Col>
                        <Col>
                            <AddCartButton productId={id} />
                        </Col>
                    </Row>
                </Card.Body>
            </Card>
        </Col>
    )
}
ProductItem.propTypes = {
    id: PropTypes.number.isRequired,
    price: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired
}