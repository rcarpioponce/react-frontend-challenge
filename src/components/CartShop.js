import React, { useContext } from 'react'
import { Row, Col, Badge, Popover, Button, OverlayTrigger } from 'react-bootstrap';
import { Cart2 } from 'react-bootstrap-icons';

import Context from '../context/Context';
import { CartItem } from './CartItem';


export const CartShop = () => {
    const [context, setContext] = useContext(Context)
    const countProducts = context.length

    let totalProducts = 0;

    context.map( product => (
      totalProducts+=product.price
    ));

    const popover = (
        <Popover id="popover-basic">
          <Popover.Title as="h3">Cart Detail</Popover.Title>
          <Popover.Content>
              <Row>
                  <Col>Item</Col>
                  <Col>Price</Col>
              </Row>
            {
                context.map( product => (
                    <CartItem {...product} key={product.id} />
                ))
            }
            <hr />
            <Row>
              <Col>Total:</Col>
              <Col>$ {totalProducts}</Col>
            </Row>
          </Popover.Content>
        </Popover>
      );
      
      const DetailCart = () => (
        <OverlayTrigger trigger="click" placement="left" overlay={popover}>
          <Button variant="success">
            <Cart2 color="white" />
            <Badge bg="primary">{ countProducts }</Badge>
          </Button>
        </OverlayTrigger>
      );

    return (
        <Row>
            <Col className='text-right'>
                <DetailCart />                
            </Col>
        </Row>
    )
}
