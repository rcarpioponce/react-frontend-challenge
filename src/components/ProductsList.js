import React from 'react'
import { Col, Row, Spinner } from 'react-bootstrap'
import { ProductItem } from './ProductItem';
import { useProductsFetch } from '../hooks/useProductsFetch';

export const ProductList = () => {

    const {data: products, loading } = useProductsFetch();
    
    return (
        <Row>
            <Col>
                <Row>
                    { 
                        loading &&
                        <Col>
                            <Spinner animation="border" role="status">
                            <span className="sr-only">Loading...</span>
                            </Spinner>  
                        </Col>
                    }
                </Row>
                <Row>
                    {
                    products.map( product  => (
                        <ProductItem 
                            key = { product.id }
                            { ...product }
                        />
                    ))}
                </Row>
            </Col>
        </Row>
    )
}
