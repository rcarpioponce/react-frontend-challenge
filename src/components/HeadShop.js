import React from 'react'
import { Row, Col } from 'react-bootstrap'
import { CartShop } from './CartShop'
export const HeadShop = () => {
    return (
        <Row>
            <Col>
                <h2>Products</h2>
            </Col> 
            <Col>
                <CartShop />
            </Col> 
        </Row>
    )
}
