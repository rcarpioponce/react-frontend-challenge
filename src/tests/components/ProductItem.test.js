import React from 'react';
import { shallow } from 'enzyme'
import { ProductItem } from "../../components/ProductItem"

describe('Pruebas en <GifGridItem />', () => {
    const id = 1;
    const price = 12.20;
    const title = 'UN TITULO';
    const image = 'https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg';
    const wrapper = shallow( <ProductItem
        id = { id } 
        price = { price }
        title={ title } 
        image={ image }  
    />)

    test('debe de mostrar el componente correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    })
})