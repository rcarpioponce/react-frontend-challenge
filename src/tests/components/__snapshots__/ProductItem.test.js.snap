// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGridItem /> debe de mostrar el componente correctamente 1`] = `
<Col
  className="mb-4 h-100"
  lg="3"
  md="3"
  xs="12"
>
  <Card
    body={false}
  >
    <CardImg
      alt="UN TITULO"
      className="img-fluid"
      src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
      variant="top"
    />
    <CardBody>
      <CardTitle>
        <Link
          to="/product/1"
        >
          UN TITULO
        </Link>
      </CardTitle>
      <Row
        noGutters={false}
      >
        <Col>
          $. 
          12.2
        </Col>
        <Col>
          <AddCartButton
            productId={1}
          />
        </Col>
      </Row>
    </CardBody>
  </Card>
</Col>
`;
